File: c/src/telegram.c
#include "../include/af_telegram.h"

#include <stdio.h>
#include <string.h>

void af_telegram_init(af_telegram_t *tg, const char *bot_token, const char *chat_id)
{
    memset(tg, 0, sizeof(*tg));
    if (bot_token && bot_token[0]) {
        strncpy(tg->bot_token, bot_token, sizeof(tg->bot_token) - 1);
        tg->bot_token[sizeof(tg->bot_token) - 1] = '\0';
    }
    if (chat_id && chat_id[0]) {
        strncpy(tg->chat_id, chat_id, sizeof(tg->chat_id) - 1);
        tg->chat_id[sizeof(tg->chat_id) - 1] = '\0';
    }
    tg->enabled = (bot_token && bot_token[0] && chat_id && chat_id[0]) ? 1 : 0;
}

int af_telegram_send(const af_telegram_t *tg, const char *message)
{
    if (!tg || !tg->enabled) return 0;

    /* Sanitise message: replace " with ' and newlines with space */
    char safe[1024];
    size_t si = 0, mi = 0;
    while (message[mi] && si < sizeof(safe) - 1) {
        char c = message[mi++];
        if (c == '"') {
            safe[si++] = '\'';
        } else if (c == '\n' || c == '\r') {
            safe[si++] = ' ';
        } else {
            safe[si++] = c;
        }
    }
    safe[si] = '\0';

    /* Build and execute curl command */
    char cmd[2048];
    snprintf(cmd, sizeof(cmd),
        "curl -s -X POST https://api.telegram.org/bot%s/sendMessage "
        "-d chat_id=%s -d text=\"%s\" > /dev/null 2>&1",
        tg->bot_token, tg->chat_id, safe);

    FILE *fp = popen(cmd, "r");
    if (!fp) return 0;
    pclose(fp);
    return 1;
}

int af_telegram_alert(const af_telegram_t *tg, const char *level, const char *msg)
{
    if (!tg || !tg->enabled) return 0;

    char formatted[1024];
    snprintf(formatted, sizeof(formatted), "[%s] %s", level ? level : "ALERT", msg ? msg : "");
    return af_telegram_send(tg, formatted);
}

File: js/src/telegram.js
/**
 * Telegram alerting — Round 11 (24/7 Ops).
 * Uses the Node.js built-in `https` module only — no external dependencies.
 * Mirrors the Python TelegramAlerter contract.
 */

import https from 'node:https';

export class TelegramAlerter {
  /**
   * @param {string} [botToken]  Bot API token; falls back to TELEGRAM_TOKEN env var
   * @param {string} [chatId]   Destination chat ID; falls back to TELEGRAM_CHAT_ID env var
   */
  constructor(botToken = '', chatId = '') {
    this._token   = botToken   || process.env.TELEGRAM_TOKEN    || '';
    this._chatId  = chatId     || process.env.TELEGRAM_CHAT_ID  || '';
    this._enabled = Boolean(this._token && this._chatId);
  }

  /** True when both token and chat-id are configured. */
  get enabled() { return this._enabled; }

  /**
   * Send a raw text message.
   * @param {string} message
   * @returns {Promise<boolean>}  Resolves to true on HTTP 200, false otherwise
   */
  send(message) {
    if (!this._enabled) return Promise.resolve(false);

    const payload = JSON.stringify({ chat_id: this._chatId, text: message });
    const url     = `https://api.telegram.org/bot${this._token}/sendMessage`;

    return new Promise((resolve) => {
      const req = https.request(url, {
        method:  'POST',
        headers: {
          'Content-Type':   'application/json',
          'Content-Length': Buffer.byteLength(payload),
        },
      }, (res) => {
        res.resume();                       // drain response body
        resolve(res.statusCode === 200);
      });
      req.on('error', () => resolve(false));
      req.write(payload);
      req.end();
    });
  }

  /**
   * Send a prefixed level alert.
   * @param {string} level  e.g. 'INFO', 'WARN', 'ERROR'
   * @param {string} msg
   * @returns {Promise<boolean>}
   */
  alert(level, msg) {
    return this.send(`[${level.toUpperCase()}] ${msg}`);
  }

  /**
   * Notify that the DrawdownGuard has transitioned state.
   * @param {string} stateName  'GREEN' | 'YELLOW' | 'RED'
   * @param {string} reason     Human-readable explanation
   * @returns {Promise<boolean>}
   */
  alertGuardState(stateName, reason) {
    const emoji = { GREEN: '✅', YELLOW: '⚠️', RED: '🚨' }[stateName.toUpperCase()] ?? '❓';
    return this.send(`${emoji} AlgoForge Guard: ${stateName}\n${reason}`);
  }

  /**
   * Notify that a trade has been placed.
   * @param {string} symbol
   * @param {string} direction  'LONG' | 'SHORT'
   * @param {number} lots
   * @param {number} entry
   * @param {number} sl        Stop-loss price
   * @param {number} tp        Take-profit price
   * @returns {Promise<boolean>}
   */
  alertTrade(symbol, direction, lots, entry, sl, tp) {
    const arrow = direction.toUpperCase() === 'LONG' ? '📈' : '📉';
    const msg   =
      `${arrow} Trade Placed\n` +
      `Symbol: ${symbol}\n` +
      `Direction: ${direction}\n` +
      `Lots: ${lots.toFixed(2)}\n` +
      `Entry: ${entry.toFixed(5)}  SL: ${sl.toFixed(5)}  TP: ${tp.toFixed(5)}`;
    return this.send(msg);
  }
}
